### ISSUE TITLE
Create option to control data type for tensor parallel all-reduce

### ISSUE THREAD

**Issue by veritas9872** [NVIDIA/TransformerEngine#1761]
**Is your feature request related to a problem? Please describe.**

I have found that using BF16 for Tensor Parallel all-reduce causes major differences in the logit values when using different TP values. This is because using BF16 buffers for TP transfers, while reducing the memory, causes errors to accumulate with each layer.

However, this can cause lots of problems when logit values are important. Currently, an average error of 15% is not uncommon when using different TP levels with BF16 buffers. This error is negligible when using FP32 buffers.

Also, now that async-TP has been introduced, I think that using FP32 for TP data transfer will not have the same effect on latency as before.

**Describe the solution you'd like**

Many users may prefer using BF16 for TP data transfer to reduce latency. However, having an option for FP32 TP/CP model parallel would be great for accuracy sensitive users. Also, this would allow easier changing of TP degree during training, etc.

**Describe alternatives you've considered**

N/A.

**Additional context**

From the code in the TE PyTorch Linear module, I am not 100% certain if BF16 is being used for TP/CP data transfer. However, I am reasonably certain that this is the case since no documentation has been provided on the issue.

**Comment by veritas9872** [comment]
Output logits for Llama 3.2 1B Instruct using NeMo 2.

### Results for TP=1

HuggingFace outputs at the top, NeMo 2 outputs at the bottom.

![Image](https://github.com/user-attachments/assets/f5263a2c-4d18-434a-9c93-0de9ebe2f9cc)

### Results for TP=8

NeMo 2 outputs with TP=8.

![Image](https://github.com/user-attachments/assets/c8c07daf-baa2-4033-be24-21a16e2c1e66)


```python
import os

import torch
import torch.distributed as dist
from nemo.collections import llm
from nemo import lightning as nl
from nemo.lightning import io
from nemo.lightning.ckpt_utils import ckpt_to_context_subdir
from transformers import AutoModelForCausalLM, AutoTokenizer


def convert():
    model_name = "meta-llama/Llama-3.2-1B-Instruct"
    nemo_model_path = "nemo_llama_32_1b_instruct"

    llm.import_ckpt(
        model=llm.LlamaModel(llm.Llama32Config1B()),
        source=f"hf://{model_name}",
        output_path=nemo_model_path,
    )


@torch.inference_mode()
def check(text: str):
    model_name = "meta-llama/Llama-3.2-1B-Instruct"
    nemo_model_path = "nemo_llama_32_1b_instruct"

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    input_ids = tokenizer(
        text=text,
        add_special_tokens=False,
        return_attention_mask=False,
        return_tensors="pt",
    ).input_ids

    device = torch.device("cuda:2")
    with torch.cuda.device(device):
        hf_model = AutoModelForCausalLM.from_pretrained(
            model_name,
            torch_dtype=torch.float32,
        )
        hf_model.tie_weights()
    hf_model.config.use_cache = False
    hf_model.eval()
    hf_model.to(device)

    strategy = nl.MegatronStrategy(
        tensor_model_parallel_size=1,
    )
    trainer = nl.Trainer(
        devices=[1],
        accelerator="gpu",
        strategy=strategy,
        plugins=nl.MegatronMixedPrecision(precision="bf16-mixed"),
    )
    trainer.strategy._setup_optimizers = False
    fabric = trainer.to_fabric()
    model: io.TrainerContext = io.load_context(
        path=ckpt_to_context_subdir(nemo_model_path),
        subpath="model",
    )
    model = fabric.load_model(nemo_model_path, model).cuda(1)
    model = model.module
    model.eval()

    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        hf_output = hf_model(input_ids=input_ids.to(device)).logits

    position_ids = torch.arange(input_ids.size(1), dtype=torch.int64, device=1)
    nemo_output = model(
        input_ids=input_ids.cuda(1),
        attention_mask=None,
        position_ids=position_ids,
    )

    breakpoint()


@torch.inference_mode()
def check_tp(text: str):
    dist.init_device_mesh(
        device_type="cuda",
        mesh_shape=(8,),
        mesh_dim_names=("tp",),
    )
    local_rank = int(os.getenv("LOCAL_RANK", 0))
    model_name = "meta-llama/Llama-3.2-1B-Instruct"
    nemo_model_path = "nemo_llama_32_1b_instruct"

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    input_ids = tokenizer(
        text=text,
        add_special_tokens=False,
        return_attention_mask=False,
        return_tensors="pt",
    ).input_ids

    strategy = nl.MegatronStrategy(
        tensor_model_parallel_size=8,
    )
    trainer = nl.Trainer(
        devices=8,
        accelerator="gpu",
        strategy=strategy,
        plugins=nl.MegatronMixedPrecision(precision="bf16-mixed"),
    )
    trainer.strategy._setup_optimizers = False
    fabric = trainer.to_fabric()
    model: io.TrainerContext = io.load_context(
        path=ckpt_to_context_subdir(nemo_model_path),
        subpath="model",
    )
    model = fabric.load_model(nemo_model_path, model).cuda(local_rank)
    model = model.module
    model.eval()

    position_ids = torch.arange(
        input_ids.size(1),
        dtype=torch.int64,
        device=local_rank,
    )
    nemo_output = model(
        input_ids=input_ids.cuda(local_rank),
        attention_mask=None,
        position_ids=position_ids,
    )

    if local_rank == 0:
        breakpoint()
    else:
        dist.barrier()
```

**Comment by veritas9872** [comment]
Commands:

```bash
python -m fire debug.py convert
```

```bash
python -m fire debug.py check --text "Summer is warm. Winter is cold."
```

```bash
torchrun --nproc-per-node=8 --nnodes=1 --standalone -m fire debug.py check_tp --text "Summer is warm. Winter is cold."
```

**Comment by veritas9872** [comment]
I used the NeMo 25.04 NGC container for my experiments.
The prompt "Summer is warm. Winter is cold." was chosen because it is known to cause massive activations (https://openreview.net/forum?id=F7aAhfitX6#discussion).

**Comment by 0z5a** [comment]
@sbhavani @veritas9872 I’d like to take a scoped version of this.

I noticed the TP-invariant work in #2977 explicitly dropped NVTE_FP32_TP_REDUCE as a separable feature, so I’d like to handle that reduction-dtype piece independently rather than mix it into TP-invariant GEMM.

My initial scope would be the plain, non-overlapped TP all-reduce path:

* keep the current reduction dtype as the default;
* add an opt-in FP32 reduction path;
* separate local GEMM numerical differences from collective-reduction error in the tests;
* compare TP=2/4 BF16 vs. FP32 reduction on multi-L20;
* report accuracy, latency, temporary-memory, and communication-volume impact.

I’d keep Userbuffers / overlap / async handling out of the first PR unless there is a preferred existing abstraction that should cover them from the start.

If this scope looks reasonable, I’d be happy to take it.

**Comment by 0z5a** [comment]
Follow-up on the scoped version I proposed: I have the measurements now. Summary — the option does what the issue asks numerically, and on this shape it costs ~1.9× in the synchronous path.

**Environment**: 2× L20 (cc 8.9, full peer access), TE `main` @ `9233c73a` built with `NVTE_CUDA_ARCHS=89 NVTE_WITH_NCCL_EP=0`, torch 2.11.0+cu128, CUDA 12.8, NCCL 2.28.9. Row-parallel `te.Linear`, `parallel_mode="row"`, `tp_size=2`, `sequence_parallel=False`, `ub_overlap_rs_fprop=False` — i.e. the plain synchronous non-Userbuffers path, confirmed from the module itself.

**1. Collective rounding, separated from local GEMM rounding**

Following the oracle split from my earlier comment (FP64 sum over the exact quantized BF16 values that enter the collective, so local GEMM error is excluded):

| case | BF16 collective | FP32 collective |
|---|---|---|
| magnitudes spanning 1e6 (1.0 … 999424) | 999425 → **999424**, abs err **1.0** | abs err **0.0** |
| uniform magnitudes | abs err 0.0 | exact |

That is the case the issue is about: with a wide magnitude range the BF16 accumulator drops the small term entirely, and the FP32 reduction recovers it.

**2. Full row-parallel forward (tp=2, M=8192, K=N=4096, BF16)**

Compared against an FP64 reference computed from the same feature-shard:

* default BF16 reduce: max abs err `0.0625`
* opt-in FP32 reduce: max abs err `0.0625`
* **bitwise identical to the default** (`mismatch_elems = 0`)

That last line is expected rather than disappointing: with tp=2 there is a single final rounding back to BF16, so widening the accumulator does not change the answer on this shape. I am reporting it because it means the option cannot be justified by "it is more accurate" on every shape — the gain only appears when the number of accumulation terms or the dynamic range is large enough. Worth confirming whether you want a tp=4 number as well; my tp=4 run did not happen (see below).

**3. Cost (tp=2, M=8192, K=N=4096)**

| path | median | min |
|---|---|---|
| default BF16 reduce | 20.10 ms | 9.92 ms |
| opt-in FP32 reduce | 38.24 ms | 19.56 ms |
| ratio | 1.90× | 1.97× |

Caveat I want to be explicit about: the host was heavily shared during this run (load average 720–4400), and the median is well above the min, so the honest read is "≈2× on this shape, measured under contention". The extra cost is the FP32 staging buffer, the cast, and the mandatory synchronisation — the option refuses `async_op=True` rather than pretending to be async, since the cast-back cannot happen before the collective completes.

**Scope I implemented**

* `distributed.allreduce()` gains an optional `reduction_dtype`; only `torch.float32` is accepted, the collective runs over a float32 copy and the result is rounded back into the caller's tensor so the existing return contract (same object / shape / device / dtype) is preserved.
* the parameter is threaded to the row-parallel forward only; everything else keeps its dtype.
* `Linear.forward()` validates up front and **raises** for combinations it does not cover (non-row-parallel, `tp_size == 1`, `sequence_parallel`, Userbuffers) rather than silently reducing in BF16.
* default `None` adds no allocation, cast, or extra sync.

**Not covered / not verified**

* backward all-reduce, sequence-parallel reduce-scatter, and the Userbuffers/overlap paths are untouched — this is forward-only, as scoped.
* no bitwise guarantee across TP degree or architecture.
* **tp=4 did not run**: I could not get 4 L20s simultaneously free on the shared box during this window. The TP-invariant separation against #2977 is also untouched here.

If the forward-only shape of this is not what you want, tell me and I will re-scope before sending a PR. I have not opened one yet.

**Comment by 0z5a** [comment]
Opened #3555 for the opt-in FP32 row-parallel reduction. On a 2-rank row-parallel forward the BF16 reduction differs from exact by ~1.0 while FP32 reproduces it to 0.0; the widening costs ~1.90× the reduction time, so it stays opt-in per call with the default untouched, and unsupported combinations are refused rather than silently reduced in BF16.
